`rust-analyzer` discoverConfig integration

bobozaur opened the pull request:
Adds a target that can be used for project auto-discovery by using the `discoverConfig` settings as described in the `rust-analyzer` user manual.

Unlike the `gen_rust_project` target, this can be used for dynamic project discovery, and passing `{arg}` to `discoverConfig.command` can split big repositories into multiple, smaller workspaces that `rust-analyzer` switches between as needed. Large repositories can make it OOM.

At amo, we've used a similar implementation for a while with great success, which is why we figured we might upstream it. The changes also include two additional output groups to ensure that proc-macros and build script targets are built, as `rust-analyzer` depends on these to provide complete IDE support.

Additionally, the PR makes use of the `output_base` value in `bazel` invocations. We found it helpful to have tools such as `rust-analyzer` and `clippy` run on a separate bazel server than the one used for building. And a `config_group` argument was added to provide the ability to provide a config group to `bazel` invocations.

An attempt to get codelens actions to work was done as well, particularly around tests and binaries. They seem to work, but I'm not 100% sure whether the approach taken is the right one.

Closes #2755 .

bobozaur commented:
@sam-mccall I noticed you've been working on things adjacent to this in other recent PRs. Just adding you here for visibility reasons. Any advice you can offer regarding the PR is more than welcome.

sam-mccall commented:
> @sam-mccall I noticed you've been working on things adjacent to this in other recent PRs. Just adding you here for visibility reasons. Any advice you can offer regarding the PR is more than welcome.

Thanks! Yes I had a draft version running locally that I'd planned to clean up and send, but I'm happy you got there first.
I'm not an owner here, but happy to provide feedback and test this out.

This PR combines a few logically-separate changes. I get it - there are a bunch of details that all need to be right for this to work end-to-end. (I ran into some overlapping-but-different set of these myself). I'll try to understand them all and provide high-level comments first, and then let you figure out whether to split them out and land the more "obvious" stuff first. (e.g. having a separate output-base and injecting blaze configuration are useful, but different users will have different needs here).

(I did spot this last friday and started to review, but have been unwell this week...)

sam-mccall reviewed (commented):
This is great stuff, really thorough job! Some parts of the protocol (logging & runnables) were totally new to me.

High level:

- starlark changes (proc macros & build scripts) are good. Related: #3031 includes some other generated files.
- A separate binary for `discover` sounds good to me (I had this as a flag on `gen_rust_project`, but it's messy)
- this is background infra that should "do what I mean", as such we want fewer flags and more detection.
- some things we will eventually want should inform flags/detection:
  - become a standalone binary (not invoked through `blaze run`)
  - automatically select an `output_base`
- a `BlazeSpec` (argv0, workspace, execroot, ...) would be a useful abstraction, replacing `Config` and long param lists
- clearer split between main (config + protocol) and library (build+query+describe) would make the code easier to follow.

Let me know what you think - as I said I'm not an owner here. I can nag one to get you a stamp :-) but also feel free to get a second opinion instead.

If you're interested, this could be a few PRs - might go quicker, but totally up to you:

- proc macros + build scripts
- refactor current bin+lib to make lib reusable
- add discover tool
- runnables

bobozaur commented on rust/private/rust_analyzer.bzl:
```diff
@@ -142,23 +161,29 @@ def _rust_analyzer_aspect_impl(target, ctx):
     return [
         rust_analyzer_info,
-        OutputGroupInfo(rust_analyzer_crate_spec = rust_analyzer_info.crate_specs),
+        OutputGroupInfo(
```
I'm on board with having out dirs be a part of `rust_analyzer_srcs` but proc-macros might be worth having as a separate output group.

My reasoning is that we like having the whole setup ready when `rust-analyzer` starts, runs a build and does its indexing. And to get full code support proc-macros must be already compiled.

In my opinion the best course of action, or at least the one we could start with, is to have proc-macros separate and perhaps have some config flag to enable/disable them.

What do you think?

bobozaur replied:
Renamed `rust_analyzer_build_info_out_dirs` to `rust_analyzer_srcs`.

bobozaur replied:
Adjusted

bobozaur commented on rust/private/providers.bzl:
```diff
@@ -162,17 +162,21 @@ RustAnalyzerInfo = provider(
         "env": "Dict[String: String]: Environment variables, used for the `env!` macro",
         "id": "String: Arbitrary unique ID for this crate",
-        "proc_macro_dylib_path": "File: compiled shared library output of proc-macro rule",
+        "proc_macro_dylib": "File: compiled shared library output of proc-macro rule",
```
This is related to having proc-macro crates built by `rust-analyzer` when it starts so that it can provide full code support.

If proc-macros are not found then `rust-analyzer` complains about that and provides errors. So the get everything in top shape, the proc-macro dependencies of a target must be compiled. And to get them compiled, the proc-macro target outputs get their own output group. 

So I changed this here from only retaining the path of the proc-macro dylib to the dylib file itself, which can then be exported through the output group.

bobozaur replied:
Updated the comment with your suggestion.

bobozaur commented on rust/private/rust_analyzer.bzl:
```diff
@@ -221,6 +247,13 @@ def _create_single_crate(ctx, attrs, info):
     crate["root_module"] = path_prefix + info.crate.root.path
     crate["source"] = {"exclude_dirs": [], "include_dirs": []}
 
+    # Store build system related info only for local crates
```
I think "local" is the right word.

I'll adjust the comment, but this is indeed to only provide build information only for crates in the source tree and not generated source files or external dependencies.

Your guess is right, though. The build info is used for runnables, but it is _also_ used for build file watching, so that `rust-analyzer` refreshes the workspace is a build file changes (similar to how it would refresh if you add a dependency to a `Cargo.toml` file).

bobozaur replied:
Adjusted the comment.

bobozaur commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,94 @@
+
+#[derive(Debug, Args)]
+struct DiscoverProjectArgs {
+    /// The build file to use as Rust workspace root when not
```
Yeah, this is a bit confusing. I'll add the zulip thread where I discussed this a bit with some `rust-analyzer` people: https://rust-lang.zulipchat.com/#narrow/channel/185405-t-compiler.2Frust-analyzer/topic/DiscoverArgument.20for.20non-cargo.20workspaces.

The main idea is that `rust-analyzer` can be configured to either always pass the argument or never pass it through the `discoverConfig` settings section.

If set to never pass the argument, then the whole source tree will be treated as a single workspace. The `DiscoverProject::Finished` enum that the `discoverCommand` returns to `rust-analyzer` must contain a buildfile as well, so in this case it would be the buildfile at the root of the workspace. I agree that we should perhaps auto-detect the `BUILD` or `BUILD.bazel` file in the current directory instead of this.

Anyway, treating everything as a single workspace is fine for smaller projects, but `rust-analyzer` will OOM on large projects, hence the ability to have `rust-analyzer` pass an argument to the `discoverCommand` so that it will essentially split the codebase into smaller workspaces. 

We currently do not use the split workspace approach because `rust-analyzer` can handle our entire codebase still and the caveat with this is that the top-level package is the root of the workspace. So if you rename something like a function or struct name, this won't propagate to dependents of the crate that reside higher the in the directory tree (package wise).

I looked again at the `rust-analyzer` code and it seems that the `DiscoverProject::Finished` buildfile field is really only used to determine the workspace root by [getting the file's parent](https://github.com/bobozaur/rust-analyzer/blob/b51f9bc736dc0472481a47d7c05de2901323e543/crates/rust-analyzer/src/config.rs#L1744). So I think that in the absence of an argument being passed from `rust-analyzer`, we can instead look for a `WORKSPACE.bazel` or `MODULE.bazel` file and use that. It does not have to be a `BUILD.bazel` file.

bobozaur replied:
Removed the `default_buildfile` argument and instead resorted to probing the fs for any of  [`MODULE.bazel`, `REPO.bazel`, `WORKSPACE.bazel`, `WORKSPACE`].

bobozaur commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,94 @@
+}
+
+fn main() {
+    // Treat logs as progress messages.
```
It does, yup. Other editors too like Zed or Neovim.

bobozaur commented on tools/rust_analyzer/rust_project.rs:
```diff
@@ -108,6 +223,31 @@ pub fn generate_rust_project(
         sysroot_src: Some(sysroot_src.into()),
         crates: Vec::new(),
+        runnables: vec![
+            Runnable {
```
> What's the intended use of build?

Don't laugh, but I have absolutely no idea. I can't find any place where it is used in `rust-analyzer` so it might be a leftover. I initially just copied this from the buck2 implementation, only refining it later. I don't see a reason to keep it, to be honest. 

But I think the best course of action would be to raise some issue with `rust-analyzer` and see if this serves any purpose before removing it, just to ensure we're not omitting something here.

bobozaur replied:
Regarding passing `--config` and other args, I'm not entirely sure either. It's a very good question so thanks for spotting this, I missed it.

From the top of my head, I'd argue that we _should not_ pass them. The main idea behind the flags are to get the tool itself running in a way that does not obstruct regular usage, like building, running or testing targets. Runnables, while invoked by the tool, do qualify as "regular usage", I think. It's just that instead of typing a command in the terminal you invoke them through the IDE code lens.

No strong opinions here though.

bobozaur replied:
Note that I ended up removing the `--config_group` CLI argument to err on the side of simplicity.

bobozaur replied:
Regarding the `RunnableKind::Check` variant, it seems that it was recently added here: https://github.com/rust-lang/rust-analyzer/commit/71a78a9cdce89b05a5d497b43fb80d4412761556. If I were to guess, this is perhaps meant to help with the implementation of `cargo check` like support.

I have no strong opinions, though. We could remove this entirely, simply comment it out or leave it in preparation for better flycheck support for non-cargo projects from `rust-analyzer`'s side.

bobozaur commented on tools/rust_analyzer/BUILD.bazel:
```diff
@@ -2,9 +2,28 @@ load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
 load("//tools:tool_utils.bzl", "aspect_repository")
 
+rust_binary(
+    name = "discover_rust_project",
```
No strong opinions here, but I simply followed the existent "convention". I'd also argue that including `bazel` in the name here is slightly awkward, since this is in itself a `bazel` target, so I personally feel like the relation is implied.

If this were to be packaged as a binary, then the `discover_bazel_rust_project` name is suitable. But renaming the target seems redundant to me.

bobozaur replied:
Renamed

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+    pub bazel: Utf8PathBuf,
+
+    /// A `--config` directive that gets passed to Bazel to be able to pass custom configurations.
+    pub config_group: Option<String>,
```
I added this because we mainly use this for two things:
- `--watchfs` to speed up some commands
- `--build_tag_filters` where we have a macro that wraps rust rules and adds a `rust_analyzer` tag.

Perhaps that this could actually be omitted entirely. The `--watchfs` thing is not that important if the `bazel query` in `query_buildfile_for_source_file` gets removed (I think it might still speed up `aquery` but it might not be worth the hassle) and the tag filter approach could be done similar to how it works for `clippy` (I think that adding a `noclippy` tag makes a target to be omitted).

Would this approach where we ignore targets that have the  `no_rust_analyzer` tag be acceptable?

bobozaur replied:
Got back to read my comment again and the more I think about it, the more I think I'm overthinking this `--config` thing. I'm not even sure of _why_ ignoring a target would be helpful as far as `rust-analyzer` is concerned. Perhaps I could simply remove the `--config` thing entirely.

What do you think?

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+    T: Args,
+{
+    // Parse the configuration flags and supplement with bazel info as needed.
+    pub fn parse_and_refine() -> anyhow::Result<Self> {
```
I guess I just wanted to relay that the function is not really the de facto `parse()` function that you get from `clap`. I ended up renaming it to `parse()` though as I agree that the name is not great.

> This could be `parse: () => Config` and `refine: Config => Config`

I have to disagree. In your example the `parse` function is simply an implementation detail that consumers should not be aware of and, in fact, never use. I personally think that callers should not have to bother with calling both `parse` and `refine`, but rather just call one function that gives them the fully featured config to use further.

Let me know if the rename seems suitable enough for you or if you have different thoughts.

bobozaur commented on rust/private/providers.bzl:
```diff
@@ -162,17 +162,21 @@ RustAnalyzerInfo = provider(
         "cfgs": "List[String]: features or other compilation `--cfg` settings",
         "crate": "CrateInfo: Crate information.",
         "crate_specs": "Depset[File]: transitive closure of OutputGroupInfo files",
+        "proc_macro_dylibs": "Depset[File]: transitive closure of OutputGroupInfo files",
```
Updated the comments.

bobozaur commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,94 @@
+use std::env;
```
Top level comment added.

bobozaur replied:
Ah, should've thought of trying that. Fixed, thanks!

bobozaur commented on tools/rust_analyzer/ra_arg.rs:
```diff
@@ -0,0 +1,129 @@
+
+    let targets = match parent_dir {
+        Some(p) if !p.as_str().is_empty() => format!("//{p}/..."),
+        _ => "//...".to_string(),
```
Great catch! Seems like the sensible thing to do. Adjusted!

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+    /// The path to the Bazel workspace directory. If not specified, uses the result of `bazel info workspace`.
+    pub workspace: Utf8PathBuf,
+
+    /// The path to the Bazel execution root. If not specified, uses the result of `bazel info execution_root`.
```
This was just moved from `gen_rust_project.rs` as is.

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+where
+    T: Args,
+{
+    /// The path to the Bazel workspace directory. If not specified, uses the result of `bazel info workspace`.
```
This was just moved from `gen_rust_project.rs` as is.

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+    /// The path to the Bazel execution root. If not specified, uses the result of `bazel info execution_root`.
+    pub execution_root: Utf8PathBuf,
+
+    /// The path to the Bazel output user root. If not specified, uses the result of `bazel info output_base`.
```
This was just moved from `gen_rust_project.rs` as is.

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+        let output = bazel_info_command
+            // Switch to the workspace directory if one was provided.
+            .current_dir(workspace.as_deref().unwrap_or(Utf8Path::new(".")))
+            .env_remove("BAZELISK_SKIP_WRAPPER")
```
Completely agree. Yet I'd handle this in a separate PR, perhaps one that addresses the other config related concerns you raised.

bobozaur replied:
I ended up addressing this too, mainly because while I split up the config so each binary has its own, I did factor out a `get_bazel_info()` function. So I added the `BazelCommand` trait which extends `std::process::Command` and helps with dealing with the duplicate code.

Maybe it's not ideal, but should at least be a step forward and should be easy to change depending on if/how we'll approach the `BazelSpec` idea.

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+    execution_root: Option<Utf8PathBuf>,
+
+    /// The path to the Bazel output user root. If not specified, uses the result of `bazel info output_base`.
+    #[clap(long, env = "OUTPUT_BASE")]
```
This was just moved from `main.rs` as is.

bobozaur commented on tools/rust_analyzer/lib.rs:
```diff
@@ -1,62 +1,153 @@
+    Ok(())
+}
+
+fn generate_rust_project(
```
Renamed `rust_project::generate_rust_project` to `rust_project::assemble_rust_project`.

> This one has an unclear scope: why do we have the blaze build step separately, and then aquery + assemble as one step when build + aquery are so tightly coupled?

This was how things were previously done. I have no strong opinions on the approach here, but I once again feel like dealing with this is outside the scope of this PR.

bobozaur commented on tools/rust_analyzer/rust_project.rs:
```diff
@@ -3,12 +3,30 @@
-use crate::aquery::CrateSpec;
+use crate::aquery::{CrateSpec, CrateType};
+
+/// The format that rust_analyzer expects as a response when automatically invoked.
```
Agreed! Moved `RustAnalyzerArg` to the `rust_project` module as well.

bobozaur commented on tools/rust_analyzer/lib.rs:
```diff
@@ -1,62 +1,153 @@
 
+/// Log formatting function that generates and writes a [`DiscoverProject::Progress`]
+/// message which `rust-analyzer` can display.
+pub fn discovery_progress(message: String) -> String {
```
Yeah, this mainly derived from the fact that `rust_project` contained the `write_rust_project` function. I moved stuff around a bit so that functions pertinent to a single binary appear in that binary alone.

bobozaur commented on tools/rust_analyzer/lib.rs:
```diff
@@ -1,62 +1,153 @@
+    let discovery_str =
+        normalize_project_string(&discovery_str, workspace, output_base, execution_root);
+
+    println!("{discovery_str}");
```
Yup, adjusted.

bobozaur commented on tools/rust_analyzer/lib.rs:
```diff
@@ -1,62 +1,153 @@
+    println!("{discovery_str}");
+}
+
 pub fn write_rust_project(
```
Agreed. Addressed this.

bobozaur commented on tools/rust_analyzer/rust_project.rs:
```diff
@@ -274,18 +430,33 @@ pub fn write_rust_project(
-        .replace("__OUTPUT_BASE__", output_base)
-        .replace("__WORKSPACE__", workspace);
+    let rust_project_content = serde_json::to_string_pretty(rust_project)?;
+    let rust_project_content = normalize_project_string(
```
Yeah this was pretty messy before too, mainly due to the `rust_project` module containing IO stuff for `gen_rust_project`. I adjusted stuff and introduced a trait `NormalizedProjectString` to address the normalization matter.

Mainly added the trait because I'm personally not on board with consumers having to be aware that they need to call `normalize_project_string` on a string they obtain somehow, as this should rather be some implementation detail.

Might not be the perfect implementation still, but I think it might be good enough for now. Let me know if this newer approach is still not satisfactory.

bobozaur commented on tools/rust_analyzer/lib.rs:
```diff
@@ -1,62 +1,153 @@
+mod aquery;
```
Agreed. Moved the library entry points to their respective binaries. The library contains lower level building blocks now.

bobozaur commented on tools/rust_analyzer/ra_arg.rs:
```diff
@@ -0,0 +1,129 @@
+
+/// `rust-analyzer` associates workspaces with buildfiles. Therefore, when it passes in a
+/// source file path, we use this function to identify the buildfile the file belongs to.
+fn query_buildfile_for_source_file(
```
Fair point. Tweaked this so that we instead probe the fs for a `BUILD.bazel` or `BUILD` file. 

I'm not sure there's any benefit to bothering with workspace detection in this case since a `bazel info` invocation might get called before this and even if it doesn't, there will be a `bazel build` and `bazel aquery` calls that will fail if we're not within a workspace.

bobozaur commented on rust/private/rust_analyzer.bzl:
```diff
@@ -45,6 +45,8 @@ def write_rust_analyzer_spec_file(ctx, attrs, owner, base_info):
         RustAnalyzerInfo: Info with the embedded spec file.
     """
     crate_spec = ctx.actions.declare_file("{}.rust_analyzer_crate_spec.json".format(owner.name))
+    proc_macro_dylibs = [base_info.proc_macro_dylib] if base_info.proc_macro_dylib else None
```
Alrighty.

I moved the logic into this function callers' body, leaving this as a pure passthrough. From what I could gather this function is only invoked in `extensions/prost/private/prost.bzl` and `rust/private/rust_analyzer.bzl`. 

Would you mind taking a look at the call sites just to double-check the logic is still sound?

bobozaur commented:
@sam-mccall My apologies for the delay. Winter break wasn't as restful as I hoped it would be so I got delayed a bit 🙃 .

Nevertheless, I addressed the bulk of your concerns. While all your ideas are great, I do feel though that some of them might be out of scope for this PR. In my opinion, the main objective here is getting a working solution for project auto-discovery. 

While some of the design decisions are questionable, I think an iterative approach might be better, considering that a few issues pertain to code that already existed or was simply moved around for the sake of DRY. In particular, the points revolving around a `BazelSpec`, the CLI args and bazel info autodetection seem to me like separate issues than what this PR tries to address. I've no problem helping with those as well, though I'd definitely be more comfortable doing that in a subsequent PR.

> A separate binary for discover sounds good to me (I had this as a flag on gen_rust_project, but it's messy)

Hah, I know what you mean. When I started out on this I attempted the same thing but, while do-able, it only led to convoluted code and pain 😂 .

> If you're interested, this could be a few PRs - might go quicker, but totally up to you

I think that multiple PR's would've been better if done from the start. However I only worked with bazel for a couple of months and I'm slightly afraid that if we'd be breaking this down now there's a risk of me missing individual pieces that all have to come together for this to work correctly. With the bulk of the work done and all being present here, I find it a easier to just march on with this single PR. If that's not acceptable, for any reason, do let me know.

> Let me know what you think - as I said I'm not an owner here. I can nag one to get you a stamp :-) but also feel free to get a second opinion instead.

Perhaps it might be valuable to get a code owner to review this as well, especially since some of your questions are related to pre-existent code.

sam-mccall commented:
> @sam-mccall My apologies for the delay. Winter break wasn't as restful as I hoped it would be so I got delayed a bit 🙃 .

No worries at all! I hope you got some time off at least :-)
 
> Nevertheless, I addressed the bulk of your concerns. While all your ideas are great, I do feel though that some of them might be out of scope for this PR. 

Yes, fair enough! I'm happy if we can find ways to reduce the scope here.

We should be mindful this isn't a small change: it's more than doubling the size of tools/rust_analyzer, and reusing code across binaries. When adding a lot of complexity, we will need to spend some effort improving design. 

---

On the command-line interface: I think if we're adding a *new* tool, it should not expose unnecessary options. That is, copying over the flags/knobs from `gen_rust_project` with the hope of removing some later, is not a reasonable way to reduce the scope of a PR even if it saves some code.

This interface is important - I think more so than the code that implements it. Once released, it will be difficult to remove features from it, as we cannot see or reason about the callers. Each option is a constraint that makes maintenance harder.

(Concrete example: `gen_rust_project` supports specifying/inferring workspace and execroot in many combinations. This makes it really awkward to add support for automatic ephemeral `--output_base`. For `discover_rust_project` this is an important feature, while the ability to pass `--execution_root` is borderline useless).

> While some of the design decisions are questionable, I think an iterative approach might be better, considering that a few issues pertain to code that already existed or was simply moved around for the sake of DRY.

DRY is library design.  If we lift some single-use code into a reusable module, now changes to each user are constrained by the other, and by the module boundary. If we want to iterate cheaply, we should duplicate code and customize it, and defer generalization until later. (I think this is the right approach for the CLI).

---

> > If you're interested, this could be a few PRs - might go quicker, but totally up to you
> 
> I think that multiple PR's would've been better if done from the start. However I only worked with bazel for a couple of months and I'm slightly afraid that if we'd be breaking this down now there's a risk of me missing individual pieces that all have to come together for this to work correctly. With the bulk of the work done and all being present here, I find it a easier to just march on with this single PR. If that's not acceptable, for any reason, do let me know.
> 
> > Let me know what you think - as I said I'm not an owner here. I can nag one to get you a stamp :-) but also feel free to get a second opinion instead.
> 
> Perhaps it might be valuable to get a code owner to review this as well, especially since some of your questions are related to pre-existent code.

Ok, summoning some owners: @UebelAndre @illicitonion @krasimirgg

- are you happy with what this PR is trying to do? (I hope so!)
- is it useful/sufficient if I continue to review? would someone prefer to take over now/soon?
- anyone have opinions/historical context on the configuration surface of gen_rust_project (flags, env vars, defaults, inference) and how much of it we should support?

bobozaur commented:
> DRY is library design. If we lift some single-use code into a reusable module, now changes to each user are constrained by the other, and by the module boundary. If we want to iterate cheaply, we should duplicate code and customize it, and defer generalization until later. (I think this is the right approach for the CLI).

@sam-mccall thanks for thinking about this. On a second thought I agree with you. I'll "duplicate" the config on each binary so we can tweak the discovery one without worrying about the impact on the `gen_rust_project` target.

bobozaur commented:
> @sam-mccall thanks for thinking about this. On a second thought I agree with you. I'll "duplicate" the config on each binary so we can tweak the discovery one without worrying about the impact on the gen_rust_project target.

I split up the configs so each binary has its own, but I did factor out a `get_bazel_info()` function (which I believe to be general enough to be "lib worthy") and added an extension trait for `std::process::Command` to help with deduplicating bazel command instantiations.

I also removed the `config_group` CLI arg for `discover_rust_project`, however another use case occurred to me and I'm really sorry for not remembering this sooner: besides `--watchfs` and the `--build_tag_filters` which, after more thought, are "useless", this also allowed passing in specific `--platforms`.

The `--platforms` matter is very important and we should figure out a way to address it. We could simply add a CLI arg fo it, and I might do that to kickstart the process, but my main rationale behind the config group was the enabling users to specify as many configuration settings as desired without bloating the CLI interface.

Curious about what other people think regarding this.

krasimirgg commented:
Thank you @bobozaur This is exciting!

As an owner, I'm happy to go over this PR and help out with the review.

On @sam-mccall 's questions:
>* are you happy with what this PR is trying to do? (I hope so!)

Yes! I think this is a great feature to have.
>* is it useful/sufficient if I continue to review? would someone prefer to take over now/soon?

Yes, I think you have a lot of good context on this and can provide some good feedback.

>* anyone have opinions/historical context on the configuration surface of gen_rust_project (flags, env vars, defaults, inference) and how much of it we should support?

Unfortunately I'm not familiar with this part of the codebase... @hlopko might have some insights. My opinions:
* it's very nice to keep the new project discovery as a separate tool.
* we should try to keep the existing gen_rust_project mechanism feature set.
* we should come up with a minimal set of supported features for the new project discovery. I think for a first version a common set of features that @bobozaur and @sam-mccall would be great (maybe with priorities; they don't all have to be added to this PR, as it might be simpler to review the addition of a single feature as a smaller follow-up PR).

It would be nice to set up some sort of a test that exercises the new workflow, since this work touches a lot of places and there are quite some nontrivial interactions that come into play to get this all to work. Maybe we can look up copying and adapting something of the pre-existing gen_rust_project tests.

bobozaur commented:
> It would be nice to set up some sort of a test that exercises the new workflow, since this work touches a lot of places and there are quite some nontrivial interactions that come into play to get this all to work. Maybe we can look up copying and adapting something of the pre-existing gen_rust_project tests.

@krasimirgg I would like to point out that the structure between `rust-project.json` and the message passed to `rust-analyzer` for autodiscovery is _the same_. Well, technically the `DiscoverProject::Finished` wraps the JSON that would otherwise be written to `rust-project.json`, but you get my point. Considering this and the fact that the `rust-project.json` payload is handled entirely on the library side, I don't really see a reason why tests that pass for `gen_rust_project` would not pass for `discover_rust_project` or vice-versa.

I'm on board with testing the `discover_rust_project` target too, but I'm just not sure what's worth testing. The only thing that comes to mind is a deserialization test to ensure that the `discover_rust_project` target only outputs `DiscoverProject` JSON messages. Should I implement this? Are there any other ideas of tests to implement?

bobozaur commented:
One other scenario popped up that might be worth considering (if not now, then in a future PR).

Since the tool _builds_ build scripts and proc macro targets to provide full support, if you are working on a proc macro crate or a build script and restart rust-analyzer, the build will fail and rust-analyzer will crash. This is, in my opinion, completely understandable.

However, if you're using the split workspace approach (passing `{arg}` to the `discoverConfig` command in rust-analyzer settings) then rust-analyzer will treat the current crate as the workspace root. It still crashes because it tries to build the crate, but the reality is that _it doesn't have to_, since there are no downstream consumers of the crate in the workspace.

So, essentially, the aspect should not output the `proc macro dylib` or `build info out dir` for the workspace root, this way not triggering a build of these targets and providing a better experience. This is definitely do-able, but I'm not sure what the right approach would be. We could have some sort of config setting where we pass the package to the aspect, and if the target it's currently observing is in that package, then omit the outputs I mentioned above.

Are there better ideas for handling this?

sam-mccall commented:
First my apologies - I had review comments on the previous revision, and then replied without sending them. (Why do we have that button?) I'm going to send them now, then go through your latest changes.

My thoughts on the topics you raised below - mostly these are "let's not try to solve it in this patch".

---

> @sam-mccall I keep thinking about this and the overall config situation. I'm wondering what you have in mind in terms of automatic ephemeral --output_base.

The default output_base would be `/tmp/rust_analyzer/<hash>`, where the hash is deterministically derived from the workspace root. This means that repeated invocations will hit the same bazel instance (until it times out), but it won't block regular user builds, and if you do work in multiple projects they won't stomp on each other.
(Ephemeral is probably the wrong word, sorry if that was confusing).

We don't need that feature in this PR. But I think it will be a good default, so until we have it I think we should refrain from offering other ways to configure the output base.

---

> Specifying platforms, config, bazelrc...

Please leave these out of this patch if you don't mind, it's already large enough.

(Probably my favorite mechanism is allowing the user to specify extra flags for the bazel invocation. Then they can pass `--bazelrc=...` for full flexibility, but also `--watchfs` directly if they only need that.)

---

> autodiscovery in polyglot workspaces where no argument is being passed to the discoverConfig command, so the tool will run on the target pattern //...

The no-argument-passed-to-discoverconfig case creates a lot of ambiguity and problems. Why do we need to support it?

---

> build scripts and proc macro targets

Workflows around this are interesting, and I don't understand them well.
Can I suggest we leave them out initially as there's clearly some risk, and then subsequently add them behind a `discover_rust_project` flag so we can see how well it works in practice?
(It's difficult to "just patch in" this PR, since the aspect changes need to be live in the codebase you're working on)

Ultimately, I think treating the directly-queried "active" package differently from its dependencies on is going to cause problems. Users edit code in multiple packages, different packages are "active" at different times, and we're going to get complex behavior based on sequencing and how rust-analyzer invalidates its cache.

---

> I'm on board with testing the discover_rust_project target too, but I'm just not sure what's worth testing.

It would be useful to have an end-to-end integration test, which creates a bazel workspace, runs the tool against a source file in it, and verifies the output is sensible.
Such tests are difficult to write cleanly. (There are some under `test/rust_analyzer`, they're not great).

I think this should wait until we're able to run the tool as a standalone binary (rather than under `blaze run`) as that is valuable and significantly changes how the test should be set up.

sam-mccall commented on tools/rust_analyzer/aquery.rs:
```diff
@@ -177,17 +202,35 @@ fn consolidate_crate_specs(crate_specs: Vec<CrateSpec>) -> anyhow::Result<BTreeS
         log::debug!("{:?}", spec);
         if let Some(existing) = consolidated_specs.get_mut(&spec.crate_id) {
             existing.deps.extend(spec.deps);
+            existing.env.extend(spec.env);
```
These changes look unrelated to this PR.

bobozaur replied:
I don't remember all the details but I know this had to do with the `OUT_DIR` env var exposed by build script targets. When specs get merged (like when having a test target for a lib crate) these also have to get propagated for `rust-analyzer` to correctly find generated files.

sam-mccall commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,265 @@
+        // borrow checker.
+        // See: <https://github.com/rust-lang/rust/issues/54663>
+        #[allow(clippy::unnecessary_unwrap)]
+        if workspace.is_some() && execution_root.is_some() && output_base.is_some() {
```
this looks like a premature optimization to me, just run `bazel info` unconditionally?

I think we'll end up needing to query the execution root ~always.
Even if not, `bazel info` is basically free compared to an actual invocation.

bobozaur replied:
Yup, we'll always call `bazel info` now that `execution_root` and `output_base` were removed.

sam-mccall commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,265 @@
+#[derive(Debug, Parser)]
+struct ConfigParser {
+    /// The path to the Bazel workspace directory. If not specified, uses the result of `bazel info workspace`.
+    #[clap(long, env = "BUILD_WORKSPACE_DIRECTORY")]
```
please remove the `workspace`, `execution_root`, and `output_base` flags: in the normal case these can all be detected, and it's not clear that command-line flags are the best way of handling other cases.

bobozaur replied:
Removed `execution_root` and `output_base`.

However, this fails to work without the `workspace` arg (which gets picked up from the env variable) because the execution root when calling `bazel run` won't be the workspace.

While packaging this might solve this, I think it's worth leaving it in for now.

sam-mccall commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,265 @@
+    #[clap(long, env = "OUTPUT_BASE")]
+    output_base: Option<Utf8PathBuf>,
+
+    /// The path to a Bazel binary.
```
please remove the `bazel` flag, unless you have a strong reason to configure it this way.

I believe it was added to `gen_rust_project` to support google's internal use ("blaze"), but in the context of autodiscovery this isn't a useful way to achieve that. We can solve it in another way later.

bobozaur replied:
May I ask what other solutions you have in mind?

I don't have a strong reason for configuring it this way, but it seems simpler to leave it in for now as the `discoverConfig` command is configurable.

If we remove it, we'll have to hardcode it.

sam-mccall commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,265 @@
+
+    /// The path to a `bazelrc` configuration file.
+    #[clap(long)]
+    bazelrc: Option<Utf8PathBuf>,
```
as mentioned in the discussion - I'd prefer to leave configurability to a future patch if that's OK, it's not clear whether --bazelrc is the best mechanism - letting the user specify bazel flags might be simpler.

sam-mccall commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,265 @@
+    bazelrc: Option<Utf8PathBuf>,
+
+    /// The argument that `rust-analyzer` can pass to the binary.
+    rust_analyzer_argument: Option<RustAnalyzerArg>,
```
as mentioned above, I think this can be required, at least initially

bobozaur replied:
I'm also answering here for visibility's sake. The explanation is in the other discussion. I strongly believe this should be supported and, in fact, the recommended way to use the tool. Users should fall back to passing the argument only if necessary.

sam-mccall commented on tools/rust_analyzer/command.rs:
```diff
@@ -0,0 +1,38 @@
+
+use camino::Utf8Path;
+
+/// Trait used for extending [`Command`] when the purpose is to invoke `bazel`
```
Can this be a plain function that returns `Command`?

(Or on a struct if you want to put `bazel`, `workspace` etc on it)

bobozaur replied:
I really liked the extension trait :(. But I guess it was a bit of an overkill, especially with the idea of having a `BazelSpec` struct.

Removed it in favor of a simple function.

bobozaur commented on rust/private/providers.bzl:
```diff
@@ -161,10 +161,12 @@ RustAnalyzerInfo = provider(
         "cfgs": "List[String]: features or other compilation `--cfg` settings",
         "crate": "CrateInfo: Crate information.",
-        "crate_specs": "Depset[File]: transitive closure of OutputGroupInfo files",
+        "crate_specs": "Depset[File]: transitive closure of OutputGroupInfo crate spec files",
```
Fair point. Fixed!

bobozaur commented on tools/rust_analyzer/config.rs:
```diff
@@ -0,0 +1,138 @@
+
+#[derive(Debug)]
+pub struct Config<T>
+where
```
I should've mentioned this as well, my bad! After thinking more about it I agree with you and the configs were already split. So each binary CLI interface can be developed independently.

bobozaur commented on tools/rust_analyzer/rust_project.rs:
```diff
@@ -2,13 +2,151 @@
+
+/// Trait used to convert a type that represents a Rust project into a string,
+/// normalizing it by replacing placeholders with the required counterparts.
+pub trait NormalizedProjectString {
```
Initially I renamed the trait, but frankly I never liked this solution because it felt off. It was merely about getting things going in a more elegant fashion than having to manually call that `normalize_project_string` function.

However, I started thinking after about  what the problem _really_ was. And I ended up simply doing the substitutions right after we read a file we're about to deserialize, which gets rid of this trait completely and applies to both `gen_rust_project` and `discover_bazel_rust_project`. Previously this was done after serializing the data and the formatting was different (`gen_rust_project` pretty prints the JSON but `rust-analyzer` expects everything in a single line) which is why the trait was introduced.

We lost the streamed deserialization though, but I've got something nice cooking for that. But that can come in a subsequent PR.

bobozaur commented on tools/rust_analyzer/rust_project.rs:
```diff
@@ -2,13 +2,151 @@
+    fn source_file_to_buildfile(file: &Utf8Path) -> anyhow::Result<Utf8PathBuf> {
+        // Skip the first element as it's always the full file path.
+        for dir in file.ancestors().skip(1) {
+            for entry in fs::read_dir(dir)? {
```
Great point! I implemented your proposal, thanks!

bobozaur commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,176 @@
+
+    let ra_arg = match rust_analyzer_argument {
+        Some(ra_arg) => ra_arg,
+        None => RustAnalyzerArg::Buildfile(find_workspace_root_file(&workspace)?),
```
I adjusted the `find_workspace_root_file` to look for build files too. However, it will still look for workspace files too, because what actually matters here is that we pass some file path who's parent is the workspace root. `rust-analyzer` doesn't care what this file is. And if no `BUILD`/`BUILD.bazel` file sits at the workspace root, then a workspace file will get picked up.

See: https://github.com/rust-lang/rust-analyzer/blob/0b68402d781955d526b80e5d479e9e47addb4075/crates/rust-analyzer/src/config.rs#L1825

bobozaur commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,176 @@
+    
+    log::info!("got rust-analyzer argument: {rust_analyzer_argument:?}");
+
+    let ra_arg = match rust_analyzer_argument {
```
We _could_ always require the argument, but I have to strongly oppose that.

The whole point of the argument is to allow `rust-analyzer` to work in huge projects that would otherwise make it OOM or be incredibly slow. But that should not be the preferred way of using the auto-discovery, as it has a fairly major downside:

> Since the repository gets split into smaller Rust workspaces, navigating the code through IDE features (like finding usage of a type/function) does not always work as expected. 
> 
> If you have crate `A` and crates `B` and `C` depend on it but you open some crate `A` file when it's not indexed (say you were previously working on unrelated crate `D`) then renaming functions/types or looking for usage will **not** show downstream usage in crates `B` or `C`, because the current workspace does not contain them.

Passing the argument should be the secondary way of using this in case the primary option (not passing the argument and letting `rust-analyzer` index everything) does not work. And it's not _that_ easy to topple `rust-analyzer`. It obviously depends on the crate sizes too, but I had it running fine on over 1500 crates. It's not a lot, but it's not something to sneeze at either.

I firmly believe we should find a way to support not passing the argument from the get-go.

bobozaur commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,176 @@
+        .target(Target::Stdout)
+        .init();
+
+    if let Err(e) = project_discovery() {
```
Agreed. Fixed!

bobozaur commented on tools/rust_analyzer/bin/discover_rust_project.rs:
```diff
@@ -0,0 +1,176 @@
+/// message which `rust-analyzer` can display.
+fn discovery_progress(message: String) -> String {
+    DiscoverProject::Progress { message }
+        .as_project_string()
```
Agreed. The trait is gone and this was adjusted accordingly.

bobozaur commented on tools/rust_analyzer/lib.rs:
```diff
@@ -1,62 +1,79 @@
+
+pub const BUILD_FILE_NAMES: &[&str] = &["BUILD.bazel", "BUILD"];
 
 pub fn generate_crate_info(
```
Agreed. Fixed!

bobozaur commented on tools/rust_analyzer/rust_project.rs:
```diff
@@ -2,13 +2,151 @@
+
+impl RustAnalyzerArg {
+    /// Consumes itself to return a build file and the targets to build.
+    pub fn query_target_details(
```
Hmm... I added them as associated methods because they were only used by the `RustAnalyzerArg` type, but on a second thought I agree with you. 

The associated methods have been moved, but I left the `into_target_details` method as implemented on `RustAnalyzerArg` as it uses `self`.

bobozaur commented:
@sam-mccall Sorry it took longer to get back on this, last week was incredibly busy. Good news though is that I'm taking a longer break from commercial work so I'll have a lot more time to focus on this.

> The default output_base would be `/tmp/rust_analyzer/<hash>`, where the hash is deterministically derived...

Ah, I misunderstood this then. Yeah, your suggestion sounds good. But I entirely agree with doing this later as this PR keeps growing.

> > Specifying platforms, config, bazelrc...
> 
> Please leave these out of this patch if you don't mind, it's already large enough.
> 
> (Probably my favorite mechanism is allowing the user to specify extra flags for the bazel invocation. Then they can pass `--bazelrc=...` for full flexibility, but also `--watchfs` directly if they only need that.)

If you strongly prefer doing this in a future PR, that's fine. But there are two points I'd like covered in this PR still:
* platforms support (less important but nice to have): it's incredibly valuable to be able to configure this as platform specific code would otherwise not be supported at all
* target filtering or some way to allow a broader wildcard like `//...` to work: this is mainly to allow support for not passing in an argument to `rust-analyzer`; in polyglot workspaces this ends up building other targets unrelated to Rust which is really bad.

I really like the extra flags idea, btw. Probably separating them in something like  `--bazel_startup_option` and `--bazel_arg` might be best as `clap` can then aggregate them into their own `Vec`, if I'm not mistaken. Would you dislike it that much if it got implemented in this PR?

> The no-argument-passed-to-discoverconfig case creates a lot of ambiguity and problems. Why do we need to support it?

Because that's the primary way the tool should be used and the way `rust-analyzer` provides better support overall. Passing the argument is meant as a way to provide some support to huge codebases that `rust-analyzer` would not be able to deal with otherwise by splitting them into smaller workspaces.

This however comes with limitations that do not exist when the whole codebase is part of the same workspace and indexed together. The main one is that `rust-analyzer` switches workspaces as you navigate the code, and if a crate is not indexed it will fetch and index it and its dependencies as a new workspace. But the _dependents_ of the crate are **not** indexed. This results in a worse developer experience when navigating and editing code through IDE features.

> > build scripts and proc macro targets
> 
> Workflows around this are interesting, and I don't understand them well...

Sure, agreed. The PR is getting big as it is and this is not _that_ important of a matter but more of a nice-to-have.

> It would be useful to have an end-to-end integration test...
>
> I think this should wait until we're able to run the tool as a standalone binary...

Alright, agreed.

alexkirsz commented:
Hi @bobozaur!

I just updated to your latest changes, from 0.55.6 to 0.57.1, swapping --config-group and --output-base for a custom --bazelrc file.

rust_analyzer.bazelrc
```
common --config=rust_tools

build --build_tag_filters=rust_analyzer
build --build_manual_tests
startup --output_base=bazel-rust-tools
```

It seems like rust-analyzer is no longer analyzing `std`. Other crates work fine, including vendored crates, but not `std`.

I'll keep investigating on my end.

bobozaur commented:
@alexkirsz The incorrect `sysroot`  explains the issue, as that's how `rust-analyzer` picks up the stdlib files. Note that the `bazelrc` file here is not really the desired direction in terms of config. I'll end up implementing this in a different way but I was first waiting to see what @sam-mccall thought regarding my latest proposal.

Not sure what changed to the sysroot path, but the PR is a bit behind so maybe that explains some differences?

alexkirsz commented:
I _think_ the sysroot path is correct. The rust analyzer toolchain repository should be registered, but its files aren't showing up in `bazel-bin`.

alexkirsz commented:
I figured out my issue: I was no longer passing the correct `output_base` to the initial bazel invocation:

I was doing:
```
bazel run @rules_rust//tools/rust_analyzer:discover_bazel_rust_project -- --bazelrc tools/rust_analyzer.bazelrc
```

When I should have been doing:
```
bazel --output_base=bazel-rust-tools run @rules_rust//tools/rust_analyzer:discover_bazel_rust_project -- --bazelrc tools/rust_analyzer.bazelrc
```

(tools/rust_analyzer.bazelrc configures a custom output_base)

I'm still thinking there's a bug somewhere as either the sysroot is placed in the wrong output_base, or it's placed in the right output_base but the script looks for it in the wrong output base.

bobozaur commented:
@sam-mccall I pushed another commit where the config is handled using the approach you described as being the most flexible. Two accumulators were introduced and can be filled by passing verbose individual flags such as:
```
--bazel_startup_option=--output_base=~/bazel --bazel_arg=--build_tag_filters=mytag --bazel_arg=--watchfs
```

I still believe that for a tool that should "do what I mean" the no `rust-analyzer` argument flow should be better/implicitly supported (right now users would need to group/tag Rust targets manually) but at least it is supported. 

I don't really see a problem with introducing a tag within `rules_rust` so that a simple `build_tag_filters` could be used by the auto-discovery tool with no argument passed, thus avoiding the need of a `bazel query` or something else more costly, but it feels like it's about time we wrap this PR up and address imperfections in separate ones. 

One other valuable thing would be the `BazelSpec` you mentioned where the `bazel info` related stuff and the `bazel` command construction can be abstracted away more nicely. Yet that's an implementation detail that can as well be addressed in isolation.

Let me know if you see any other pending issue here.

bobozaur commented:
> * Is this supposed to be instead of `rust-analyzer.linkedProjects`?

Yes, it is. In fact, `rust-analyzer.linkedProjects` seems to be in conflict with `rust-analyzer.workspace.discoverConfig`, so only one of them should be used.
 
>  * It seems to always be giving a progress message of `Building rust_analyzer_crate_spec files for ["//..."]`.

There are two "modes":
 - without `{arg}` passed the entire codebase gets indexed
 - with `{arg}` passed, the source/build file that `rust-analyzer` passes determines the workspace root

The former is to be preferred, as there are benefits from indexing the entire codebase in terms of IDE support and navigation. The latter is meant to indeed split the codebase into multiple smaller workspaces for the sake of allowing IDE support on huge codebases that `rust-analyzer` would not be able to handle all at once.

>  * Can the docs be updated with a recommended way of configuring this? The below is what I tested it with.

That seems okay, **but** what I tend to do is have a wrapper script that explicitly pipes `stderr` somewhere else. The reason for that is because `rust-analyzer` will treat _anything_ written to `stderr` as an error, and `bazel` likes to write a lot of stuff to `stderr`, even with the flags you set. 

I personally have a shell script with the following content:
```shell
#!/usr/bin/bash

bazel \
    run \
    @rules_rust//tools/rust_analyzer:discover_bazel_rust_project -- \
    --bazel_startup_option=--output_base=~/ide_bazel \
    --bazel_arg=--norun_validations \
    --bazel_arg=--watchfs \
    ${1:+"$1"} 2>/dev/null
```

And the IDE config section looks something like this:
```json
        "rust-analyzer": {
                "workspace": {
                    "discoverConfig": {
                        "command": [
                            "/home/bogdan/discover_bazel_rust_project.sh"
                        ],
                        "progressLabel": "rust_analyzer",
                        "filesToWatch": ["BUILD", "BUILD.bazel", "MODULE.bazel"]
                    }
                }
            }
```

> EDIT: Even if I pass a `file` or a `buildfile`, it still seems to build the whole workspace:

Is this a multi-crate workspace? Because you seem to pass `/home/cameron/Repos/bazel-lsp/BUILD`, and if `/home/cameron/Repos/bazel-lsp` is the workspace root (bazel-wise) then the resolved targets would indeed be `//...`.

> I'm also getting a lot of messages like this. Is this expected?
> 
> ```
> {"kind":"progress","message":"Skipping missing crate_spec file: \"/home/cameron/.cache/bazel/_bazel_cameron/0f48c60623c482f4373b4a0607a7289e/execroot/_main/bazel-out/k8-fastbuild/bin/external/rules_rust~~_crate~~~crate~crates__tower-layer-0.3.2/tower_layer.rust_analyzer_crate_spec.json\""}
> ```

It is. I just hooked in the progress message mechanism into the logger being used, so a lot of progress messages are logs that already existed. Doing it this way has the benefit of reusing the log messages spread across the code and makes things consistent in terms of debuggability and messages.

Right now the logging level is hardcoded to `debug`, but this could by all means be picked at runtime instead, like from the de facto `RUST_LOG` env var. I did not bother with that here because visibility is more valuable, especially now in the beginning, and because these are just piped to `rust-analyzer` and are only visible if you explicitly follow the `rust-analyzer` logs.

bobozaur commented:
Hey @sam-mccall @krasimirgg ,

Just wondering if there's something in particular that remains to be done in this PR. Haven't heard back from you guys in a while.

dimatosaurus commented:
Any chance to accelerate the review here (or at least not forget about it)? This PR is super useful, works great and is now core to my workflow!

UebelAndre commented:
Friendly ping @krasimirgg 😄

krasimirgg commented:
Hey @bobozaur, sorry for the long pause, now I'm back at reviewing this.

IMO this looks in a good shape. 

Let's add some documentation about using this (as-is now, with the user having to build / manually run the binary similarly to how gen_rust_project is configured.

Since the interactions between this and IDEs are non-tirival and to let us improve it incrementally in the future, it would be awesome to have some "somewhat-of-an end-to-end" test for this that simulates a small ide-like-interaction, similarly to https://github.com/bazelbuild/rules_rust/blob/main/test/rust_analyzer/rust_analyzer_test_runner.sh.

bobozaur commented:
@krasimirgg All done.

I got the PR up to speed with the latest changes. From what I could gather the most notable ones were:
- https://github.com/bazelbuild/rules_rust/pull/3192 - easily taken care of by merging `main`
- https://github.com/bazelbuild/rules_rust/pull/3317 - needed manual adjustments but hopefully everything is as close to before as possible, including the extended crate merging logic and newly added trace messages
- https://github.com/bazelbuild/rules_rust/pull/3254 - this came as a bit of a surprise, since I wanted to do the same thing here but was perceived as being too limited, which is why I went with the `bazel_args` and `bazel_startup_options` lists of arguments; nevertheless, I retrofitted the change from here so that the `config` argument gets transformed into a `bazel_arg` and treated the same way that `discover_bazel_rust_project` target treats them to maintain backwards compatibility.

I listed these changes here just so that reviewers or authors of those PRs could take another look at these parts of the code to double-check that they are in order.

Additionally, went ahead and added some documentation about how to use the `discover_bazel_rust_project` target in its current state. Might be a bit subjective but should be about enough to get the ball rolling.

Regarding testing, for the sake of brevity I tweaked `rust_analyzer_test_runner.sh` to also generate an `auto-discovery.json` file where it pipes the `stdout` of `discover_bazel_rust_project` and added an E2E test inspired by `static_and_shared_lib_test` that parses the file and performs the same actions as the original test. Not sure if this is close to what you had in mind, but should be a starting point for further tests. 

I'll just emphasize again the fact that the inner JSON structure between `gen_rust_project` and `discover_bazel_rust_project` is the same since they use the same underlying library to generate it, so it's debatable what exactly should be tested in one place and/or the other. I merely tried to get an example out there to hopefully make it easier to refine/add new tests.

krasimirgg reviewed (approved):
Looks great! Thank you very much!

purkhusid commented:
@bobozaur Have you had any luck getting `cargo check` like functionality working with the autodiscovery?
